pkg/exec/devfile.go

```diff
@@ -222,6 +224,61 @@ func ExecuteDevfileDebugActionWithoutRestart(client ExecClient, exec common.Exec
 	return nil
 }
 
+// ExecuteCompositeDevfileAction executes a given composite command in a devfile
+// The composite command may reference exec commands, composite commands, or both
+func ExecuteCompositeDevfileAction(client ExecClient, composite common.Composite, commandsMap map[string]common.DevfileCommand, compInfo adaptersCommon.ComponentInfo, show bool, machineEventLogger machineoutput.MachineEventLoggingClient) (err error) {
+	if composite.Parallel {
+		// Loop over each command and execute it in parallel
+		commandExecs := util.NewConcurrentTasks(len(composite.Commands))
+		for _, command := range composite.Commands {
+			cmd := command // needed to prevent the lambda from capturing the value
+			if devfileCommand, ok := commandsMap[strings.ToLower(cmd)]; ok {
+				commandExecs.Add(util.ConcurrentTask{ToRun: func(errChannel chan error) {
+					err := execCommandFromComposite(client, devfileCommand, commandsMap, compInfo, show, machineEventLogger)
+					if err != nil {
+						errChannel <- err
+					}
+				}})
+			} else {
+				return fmt.Errorf("command %q not found in devfile", command)
+			}
+		}
+
+		err := commandExecs.Run()
+		if err != nil {
+			return errors.Wrap(err, "command exec failed")
+		}
+
+	} else {
+		// Execute the commands in order
+		for _, command := range composite.Commands {
+			if devfileCommand, ok := commandsMap[strings.ToLower(command)]; ok {
```

Comment:
It's needed because when the command IDs are parsed by odo, they're made lower case, but because the command list in a composite command is just a simple list of strings, we need to make the conversion.

Without the ToLower, the check in the map will fail

Reply:
ok, gotcha.

But thinking out loud, do we want to update it in the parser itself? So that anyone who is using it in the future dont have to do it everytime they use it like and can be done once in the parser directly? 🤔

Reply:
I'm not sure there's a way to directly parse an array of strings as lower case, I'd leave as-is for now